File: legacy/code_r/ch01_Analysis_of_Economics_Data.R
# ch01_Analysis_of_Economics_Data.R - January 2026 For R
# Updated for online execution portability

########## OVERVIEW ##########

# R Program
# copyright C 2021 by A. Colin Cameron (original code)
# Updated: 2026 for portability
# Used for "Analysis of Economics Data: An Introduction to Econometrics"
# by A. Colin Cameron (2021)

# Data is streamed directly from GitHub - no local files needed
# Works in RStudio Cloud, Posit Cloud, and any online R environment

########## SETUP ##########

# Clear the workspace
rm(list=ls())

# Set random seed for reproducibility
set.seed(42)

# GitHub data URL
github_data_url <- "https://raw.githubusercontent.com/quarcs-lab/data-open/master/AED/"

# Create output directories (optional - for saving figures and tables locally)
images_dir <- "images"
tables_dir <- "tables"
dir.create(images_dir, showWarnings = FALSE, recursive = TRUE)
dir.create(tables_dir, showWarnings = FALSE, recursive = TRUE)

# Load required libraries (install if needed)
if (!require("haven")) install.packages("haven")
library(haven)

# Function to load data from GitHub
load_data <- function(filename) {
  url <- paste0(github_data_url, filename)
  tryCatch({
    read_dta(url)
  }, error = function(e) {
    message("Error loading from GitHub: ", e$message)
    message("Attempting to load from local directory...")
    read_dta(file.path("../data_stata", filename))
  })
}

########## DATA DESCRIPTION ##########

# House sale price for 29 houses in Central Davis in 1999
#     29 observations on 9 variables

########## 1.3 REGRESSION ANALYSIS ##########

# Read in the Stata data set from GitHub
data.HOUSE <- load_data("AED_HOUSE.DTA")

# Allow variables in database to be accessed simply by giving names
attach(data.HOUSE)

# Summarize the data set
cat("\n", strrep("=", 70), "\n")
cat("Data Summary\n")
cat(strrep("=", 70), "\n")
summary_stats <- summary(data.HOUSE)
print(summary_stats)

# Save summary statistics to table
write.csv(as.data.frame(do.call(cbind, lapply(data.HOUSE, summary))),
          file.path(tables_dir, "ch01_descriptive_stats.csv"))
cat("\nTable saved to:", file.path(tables_dir, "ch01_descriptive_stats.csv"), "\n")

# Fit the regression line
ols <- lm(price ~ size)

# Display regression results
cat("\n", strrep("=", 70), "\n")
cat("OLS Regression Results: price ~ size\n")
cat(strrep("=", 70), "\n")
print(summary(ols))

# Save regression results
sink(file.path(tables_dir, "ch01_regression_summary.txt"))
print(summary(ols))
sink()
cat("\nRegression summary saved to:", file.path(tables_dir, "ch01_regression_summary.txt"), "\n")

# Save regression coefficients
coef_table <- data.frame(
  coefficient = coef(ols),
  std_error = summary(ols)$coefficients[, "Std. Error"],
  t_value = summary(ols)$coefficients[, "t value"],
  p_value = summary(ols)$coefficients[, "Pr(>|t|)"]
)
write.csv(coef_table, file.path(tables_dir, "ch01_regression_coefficients.csv"))
cat("Coefficients saved to:", file.path(tables_dir, "ch01_regression_coefficients.csv"), "\n")

########## FIGURE 1.1: SCATTER PLOT WITH FITTED LINE ##########

# Display plot in R console
plot(size, price,
     xlab="House size (in square feet)",
     ylab="House sale price (in dollars)",
     pch=19, col="black",
     main="Figure 1.1: House Price vs Size")
abline(ols, col="blue", lwd=2)
legend("topleft",
       c("Actual", "Fitted"),
       lty=c(0, 1),
       pch=c(19, NA),
       col=c("black", "blue"),
       bty="o")

# Save Figure 1.1 to file
png(file.path(images_dir, "ch01_fig1_house_price_vs_size.png"),
    width=800, height=600)
plot(size, price,
     xlab="House size (in square feet)",
     ylab="House sale price (in dollars)",
     pch=19, col="black",
     main="Figure 1.1: House Price vs Size")
abline(ols, col="blue", lwd=2)
legend("topleft",
       c("Actual", "Fitted"),
       lty=c(0, 1),
       pch=c(19, NA),
       col=c("black", "blue"),
       bty="o")
dev.off()
cat("\nFigure saved to:", file.path(images_dir, "ch01_fig1_house_price_vs_size.png"), "\n")

########## ANALYSIS COMPLETE ##########

cat("\n", strrep("=", 70), "\n")
cat("ANALYSIS COMPLETE\n")
cat(strrep("=", 70), "\n")
cat("\nFigures saved to:", images_dir, "\n")
cat("Tables saved to:", tables_dir, "\n")

# Detach database
detach(data.HOUSE)

File: legacy/code_python/ch01_Analysis_of_Economics_Data.py
# %%
"""
ch01_Analysis_of_Economics_Data.py - January 2026 for Python

Chapter 1: REGRESSION ANALYSIS

To run you need file:
  AED_HOUSE.DTA
in the data_stata/ directory
"""

# %% ========== SETUP ==========

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import statsmodels.api as sm
from statsmodels.formula.api import ols
import random
import os

# Set random seeds for reproducibility
RANDOM_SEED = 42
random.seed(RANDOM_SEED)
np.random.seed(RANDOM_SEED)
os.environ['PYTHONHASHSEED'] = str(RANDOM_SEED)

# GitHub data URL (reliable mirror of Cameron's textbook data)
GITHUB_DATA_URL = "https://raw.githubusercontent.com/quarcs-lab/data-open/master/AED/"

# Output directories (optional - for saving figures and tables locally)
IMAGES_DIR = 'images'
TABLES_DIR = 'tables'
os.makedirs(IMAGES_DIR, exist_ok=True)
os.makedirs(TABLES_DIR, exist_ok=True)

# %% ========== OVERVIEW ==========

# This Python program does analysis for Chapter 1
#  1.3 REGRESSION ANALYSIS

# %% ========== DATA DESCRIPTION ==========

# House sale price for 29 houses in Central Davis in 1999
#     29 observations on 9 variables

# %% ========== 1.3 REGRESSION ANALYSIS ==========

# Read in the Stata data set
# Data is streamed directly from GitHub (reliable mirror)
data_house = pd.read_stata(GITHUB_DATA_URL + 'AED_HOUSE.DTA')

# %% Explore data structure

# Display summary statistics
print("=" * 70)
print("CHAPTER 1: ANALYSIS OF ECONOMICS DATA")
print("=" * 70)
print("\nData Summary:")
data_summary = data_house.describe()

# %% Calculate statistics
print(data_summary)

# Save descriptive statistics table
data_summary.to_csv(os.path.join(TABLES_DIR, 'ch01_descriptive_stats.csv'))
print(f"Table saved to: {os.path.join(TABLES_DIR, 'ch01_descriptive_stats.csv')}")

print("\nData Info:")
print(data_house.info())

# Fit the regression line: price ~ size
# Using statsmodels formula API (similar to R's lm)

# %% Estimate regression model

model = ols('price ~ size', data=data_house).fit()

# Display regression results
print("\n" + "=" * 70)
print("OLS REGRESSION RESULTS: price ~ size")
print("=" * 70)

# %% Display regression results

print(model.summary())

# Save regression summary to text file
with open(os.path.join(TABLES_DIR, 'ch01_regression_summary.txt'), 'w') as f:
    f.write(model.summary().as_text())
print(f"\nRegression summary saved to: {os.path.join(TABLES_DIR, 'ch01_regression_summary.txt')}")

# Save regression coefficients table
coef_table = pd.DataFrame({
    'coefficient': model.params,
    'std_err': model.bse,
    't_value': model.tvalues,
    'p_value': model.pvalues,
    'conf_lower': model.conf_int()[0],
    'conf_upper': model.conf_int()[1]
})
coef_table.to_csv(os.path.join(TABLES_DIR, 'ch01_regression_coefficients.csv'))
print(f"Coefficients table saved to: {os.path.join(TABLES_DIR, 'ch01_regression_coefficients.csv')}")

# %% ========== FIGURE 1.1: SCATTER PLOT WITH FITTED LINE ==========

# Create the scatter plot with fitted regression line
fig, ax = plt.subplots(figsize=(10, 6))

# Scatter plot of actual data
ax.scatter(data_house['size'], data_house['price'],
           color='black', s=50, label='Actual', alpha=0.7)

# Add fitted regression line
ax.plot(data_house['size'], model.fittedvalues,
        color='blue', linewidth=2, label='Fitted')

# Labels
ax.set_xlabel('House size (in square feet)', fontsize=12)
ax.set_ylabel('House sale price (in dollars)', fontsize=12)
# ax.set_title('Figure 1.1: House Price vs Size', fontsize=14, fontweight='bold')  # Removed: redundant with LaTeX caption
ax.legend(loc='upper left')
ax.grid(True, alpha=0.3)

# Save figure
output_file = os.path.join(IMAGES_DIR, 'ch01_fig1_house_price_vs_size.png')
plt.tight_layout()
plt.savefig(output_file, dpi=300, bbox_inches='tight')
print(f"\nFigure saved to: {output_file}")

# Display the plot
plt.show()

# %% ========== REGRESSION DIAGNOSTICS ==========

print("\n" + "=" * 70)
print("REGRESSION COEFFICIENTS")
print("=" * 70)
print(f"Intercept: ${model.params['Intercept']:,.2f}")
print(f"Slope (size coefficient): ${model.params['size']:,.2f}")
print(f"R-squared: {model.rsquared:.4f}")
print(f"Adj. R-squared: {model.rsquared_adj:.4f}")
print(f"Number of observations: {int(model.nobs)}")

# %% ========== COMPARISON WITH STATA/R OUTPUT ==========

print("\n" + "=" * 70)
print("INTERPRETATION")
print("=" * 70)
print(f"For every additional square foot of house size,")
print(f"the sale price increases by approximately ${model.params['size']:,.2f}")
print(f"\nThe model explains {model.rsquared*100:.2f}% of the variation in house prices.")

print("\n" + "=" * 70)
print("ANALYSIS COMPLETE")
print("=" * 70)
